ipa-core/src/protocol/ipa_prf/aggregation/mod.rs

```diff
@@ -0,0 +1,257 @@
+                    record_id,
+                    chunk.attributed_breakdown_key_bits,
+                    chunk.capped_attributed_trigger_value,
+                    B,
+                    false,
+                )
+                .await
+            }
+        },
+        || AttributionOutputs {
+            attributed_breakdown_key_bits: Replicated::ZERO,
+            capped_attributed_trigger_value: Replicated::ZERO,
+        },
+    )
+    .then(|fut| async move { fut.await.map(Chunk::into_raw) });
+
+    let aggregation_input = Box::pin(
+        row_contributions_stream
+            .map_ok(|chunk| {
+                // Aggregation intermediate transpose
+                Vec::transposed_from(chunk.as_slice()).unwrap_infallible()
+            })
+            .try_flatten_iters::<BitDecomposed<_>, Vec<_>>(),
+    );
+
+    aggregate_values::<_, _, B>(ctx, aggregation_input, num_chunks * N).await
+}
+
+pub type AggResult<const B: usize> = Result<BitDecomposed<Replicated<Boolean, B>>, Error>;
+
+pub async fn aggregate_values<'fut, C, OV, const B: usize>(
+    ctx: C,
+    mut aggregated_stream: Pin<Box<dyn Stream<Item = AggResult<B>> + Send + 'fut>>,
+    mut num_rows: usize,
+) -> Result<Vec<Replicated<OV>>, Error>
+where
+    C: UpgradedContext<Boolean, Share = Replicated<Boolean>>,
+    OV: SharedValue + U128Conversions + CustomArray<Element = Boolean>,
+    Boolean: FieldSimd<B>,
+    Replicated<Boolean, B>: BooleanProtocols<C, Boolean, B>,
+    Vec<Replicated<OV>>:
+        for<'a> TransposeFrom<&'a BitDecomposed<Replicated<Boolean, B>>, Error = LengthError>,
+{
+    let mut depth = 0;
+    while num_rows > 1 {
+        // We reduce pairwise, passing through the odd record at the end if there is one, so the
+        // record count gets rounded down, but the number of outputs gets rounded up.
+        let par_agg_ctx = ctx
+            .narrow(&Step::Aggregate(depth))
+            .set_total_records(TotalRecords::Indeterminate);
+        num_rows = (num_rows + 1) / 2;
+        aggregated_stream = Box::pin(aggregated_stream.try_chunks(2).enumerate().then(
+            move |(i, chunk_res)| {
+                let ctx = par_agg_ctx.clone();
+                async move {
+                    match chunk_res {
+                        Err(e) => Err(e.1),
+                        Ok(mut chunk_vec) if chunk_vec.len() == 1 => Ok(chunk_vec.pop().unwrap()),
+                        Ok(mut chunk_pair) => {
+                            assert_eq!(chunk_pair.len(), 2);
+                            let b = chunk_pair.pop().unwrap();
+                            let a = chunk_pair.pop().unwrap();
+                            let (mut sum, carry) =
+                                integer_add::<_, _, _, _, B>(ctx, RecordId::from(i), &a, &b)
+                                    .await?;
+                            if a.len() < usize::try_from(OV::BITS).unwrap() {
+                                sum.push(carry);
+                            }
```

Comment:
So this implements **wrapping addition**, basically addition modulo `2^OV::BITS`. This could potentially yield unexpected results (like zero) when the sum wraps around.

An alternative would be to implement a "saturating addition".

We could implement that with:

```
if /* it could potentially overflow on this step */ {
  let sum = integer_sat_add(/* stuff */)
} else {
  let (mut sum, carry) = integer_add(/* stuff */).await?;
  sum.push(carry);
}
```

Reply:
do we want to move this circuit to a separate function? Seeing `wrapping_integer_add` or `sat_integer_add` here makes the intent more obvious imo

Reply:
The old version didn't have saturating math either, but I agree that now that we can do it relatively cheaply, it's probably worth doing. Is there any interaction with DP?

Reply:
The reason we never had such logic before is that we were moving into a huge field, and assumed we would never encounter a sum that reached 2^32. Now that we wrap around at a configurable value, it's possible to select a number that's fairly low.

I think it might be fine to just document this behavior and warn everyone to please select an output type that's sufficiently large so that there is no overflow. That would keep the code simpler. I don't think there's any impact on DP. If there is wrapping, it will essentially *decrease* the overall contribution of some people to the query output. But it's not predictable. Some people will get their contribution removed and others will not. So if we take a "worst case" scenario type of approach, we probably can't say anything definitive and cannot benefit from this.